plugins/autoid/plugin.js

```diff
@@ -158,6 +160,208 @@
         return !(/\w/.test(character));
       }
 
+      // This modifies the dialog for the 'link' plugin to allow the option of
+      // internal document linking to autoid headings. Original 'link' plugin
+      // code can be found at:
+      // https://github.com/ckeditor/ckeditor-dev/tree/master/plugins/link
+
+      function modifyLinkDialog(ev) {
+        var dialog = ev.data;
+
+        if (dialog.name == 'link') {
+          var def = ev.data.definition,
+            infoTab = def.contents[0],
+            linkTypeTab = infoTab.elements[1];
+
+          // Add option for heading links to linkType dropdown
+          linkTypeTab.items.push(['Link to heading in the text', 'heading']);
+
+          // Add additional content to 'Link Info' tab for heading links
+          infoTab.elements.push(createHeadingLinkContent());
+
+          // Modify linkType's 'onChange' function to accomodate new heading option
+          linkTypeTab.onChange = modifiedLinkTypeChanged;
+
+          // Modify main dialog 'onOk' function to handle heading data
+          def.onOk = modifiedOnOk;
+        }
+      }
+
+      function modifiedLinkTypeChanged() {
```

Comment:
This looks very much like https://github.com/ckeditor/ckeditor-dev/blob/master/plugins/link/dialogs/link.js#L46 with an extra option added (because it is).

I was going to suggest trying to do this in a way where we just have a separate function that shows the new tab + calls the old function, but my 5 minute glance says I don't think this is possible without heavy refactoring of the original dialog.

Did you look into this as well?